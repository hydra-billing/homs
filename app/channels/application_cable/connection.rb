module ApplicationCable
  class Connection < ActionCable::Connection::Base
    identified_by :current_user

    def connect
      self.current_user = find_verified_user
    end

    private

    def find_verified_user
      if env['warden'].user.nil?
        reject_unauthorized_connection
      else
        env['warden'].user
      end
    end
  end
end
