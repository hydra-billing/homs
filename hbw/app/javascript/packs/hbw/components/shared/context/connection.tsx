import React, { createContext, ComponentType, FC } from 'react';
import Connection from '../../../connection';

export type ContextType = {
  options: {
    host: string,
    path: string
  },
  request: Function,
  serverURL: string
}

const ConnectionContext = createContext<ContextType | null>(null);

type InitParams = {
  connection: Connection
}

export const withConnectionContext = (
  { connection }: InitParams
) => (WrappedComponent: ComponentType) => {
  const ConnectionProvider: FC = props => (
      <ConnectionContext.Provider value={connection} >
        <WrappedComponent {...props} />
      </ConnectionContext.Provider >
  );

  return ConnectionProvider;
};

export default ConnectionContext;
