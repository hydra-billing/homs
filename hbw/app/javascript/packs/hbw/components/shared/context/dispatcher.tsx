import React, { createContext, ComponentType, FC } from 'react';
import Dispatcher from '../../../dispatcher';

export type ContextType = Dispatcher;

const DispatcherContext = createContext<Dispatcher | null>(null);

type InitParams = { dispatcher: Dispatcher };

export const withDispatcherContext = (
  { dispatcher }: InitParams
) => (WrappedComponent: ComponentType) => {
  const DispatcherProvider: FC = props => (
    <DispatcherContext.Provider value={dispatcher} >
      <WrappedComponent {...props} />
    </DispatcherContext.Provider >
  );

  return DispatcherProvider;
};

export default DispatcherContext;
