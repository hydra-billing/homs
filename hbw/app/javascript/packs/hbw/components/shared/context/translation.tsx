import React, {
  createContext, ComponentType, FC, useState, useEffect, useContext
} from 'react';
import TranslatorFactory from '../../../translator';
import { localizer } from '../../../init/date_localizer';
import ConnectionContext, { ContextType as ConnectionContextType } from './connection';

export type ContextType = {
  locale: {
    code: 'en' | 'ru' | 'es' | 'fr',
    dateTimeFormat: string
  }
  translate: (key: string, vars?: object, fallback?: string) => string,
  translateBP: (key: string, vars?: object, fallback?: string) => string,
  Localizer: {
    localizeDatetime: (str: string) => string | null,
    localizeDayMonth: (str: string) => string | null,
    localizeDayMonthYear:(str: string) => string | null
  }
}

const TranslationContext = createContext<ContextType | null>(null);

type InitParams = {
  locale: {
    code: 'en' | 'ru' | 'es' | 'fr',
    dateTimeFormat: string
  }
}

export const withTranslationContext = (
  { locale }: InitParams
) => (WrappedComponent: ComponentType) => {
  const TranslationProvider: FC = (props) => {
    const connection = useContext(ConnectionContext) as ConnectionContextType;

    const [BPTranslator, setBPTranslator] = useState({ t: () => '' });

    useEffect(() => {
      const fetchBPTranslations = async () => {
        const response = await connection.request({
          url:    `${connection.serverURL}/translations`,
          method: 'GET'
        });

        const BPTranslations = await response.json();

        setBPTranslator({ t: TranslatorFactory.getTranslatorForLocale(locale.code, BPTranslations) });
      };

      fetchBPTranslations();
    }, []);

    const translate = TranslatorFactory.getTranslatorForLocale(locale.code);
    const Localizer = localizer(locale);

    return (
      <TranslationContext.Provider value={{
        locale, translate, translateBP: BPTranslator.t, Localizer
      }}>
        <WrappedComponent {...props} />
      </TranslationContext.Provider>
    );
  };

  return TranslationProvider;
};

export default TranslationContext;
