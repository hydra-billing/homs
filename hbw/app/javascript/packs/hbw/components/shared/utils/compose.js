export default (...args) => value => args.reduceRight((acc, func) => func(acc), value);
