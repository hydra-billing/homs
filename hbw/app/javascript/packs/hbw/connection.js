import qs from 'qs';

export default class Connection {
  constructor (options) {
    this.options = options;

    if (!this.options.path) {
      this.options.path = '';
    }

    if (!this.options.host) {
      this.options.host = `${document.location.protocol}//${document.location.host}`;
    }

    this.serverURL = this.options.host + this.options.path;
  }

  request = (opts) => {
    if (opts.method.toLowerCase() === 'get') {
      const url = new URL(opts.url, window.location);

      const searchParams = [...url.searchParams]
        .reduce((obj, [key, value]) => ({ ...obj, [key]: value }), {});

      url.search = qs.stringify({
        ...searchParams,
        ...(opts.data || {}),
        payload: this.options.payload
      });

      return fetch(url.href);
    } else {
      return fetch(opts.url, {
        ...opts,
        body: JSON.stringify({
          ...opts.data,
          payload: this.options.payload
        })
      });
    }
  };
}
