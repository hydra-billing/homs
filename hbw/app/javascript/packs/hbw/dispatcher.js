class Dispatcher {
  constructor () {
    this.bind = this.bind.bind(this);
    this.unbind = this.unbind.bind(this);
    this.unbindAll = this.unbindAll.bind(this);
    this.trigger = this.trigger.bind(this);
    this.callbacksFor = this.callbacksFor.bind(this);
    this._addClient = this._addClient.bind(this);
    this._removeClient = this._removeClient.bind(this);

    this.binds = {};
  }

  bind (eventName, clientName, clbk) {
    if (!(eventName in this.binds)) {
      this.binds[eventName] = [];
    }

    this._addClient(this.binds[eventName], clientName, clbk);
  }

  unbind (eventName, clientName) {
    if (eventName in this.binds) {
      this._removeClient(this.binds[eventName], clientName);
    }
  }

  unbindAll (clientName) {
    for (let key in this.binds) {
      this._removeClient(this.binds[key], clientName);
    }
  }

  trigger (eventName, source, payload = null) {
    this.callbacksFor(eventName).forEach(callback => callback(payload, source, eventName));
  }

  callbacksFor (eventName) {
    return (this.binds[eventName] || []).map(c => c.callback);
  }

  _addClient (clients, client, clbk) {
    [...clients].forEach((c) => {
      if (c.client === client) {
        throw new Error(`Client ${client} already bound!`);
      }
    });

    clients.push({
      client,
      callback: clbk
    });
  }

  _removeClient (clients, client) {
    [...clients].forEach((c, i) => {
      if (c.client === client) {
        clients.splice(i, 1);
      }
    });
  }
}

modulejs.define('HBWDispatcher', [], () => Dispatcher);
