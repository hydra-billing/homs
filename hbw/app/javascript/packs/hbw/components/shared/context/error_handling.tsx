import Messenger from 'messenger';
import React, {
  createContext, ComponentType, FC, useState, useCallback, useContext
} from 'react';
import HBWError from '../../error';
import TranslationContext, { ContextType as TranslationContextType } from './translation';
import ConnectionContext, { ContextType as ConnectionContextType } from './connection';
import { CamundaError, WidgetErrorType } from '../utils/errors';

export type ContextType = {
  addError: Function,
  formError: React.JSX.Element,
  safeRequest: Function
}

const ErrorHandlingContext = createContext<ContextType | null>(null);

export const withErrorHandlingContext = () => (WrappedComponent: ComponentType) => {
  const ErrorHandlingProvider: FC = (props) => {
    const { translate: t } = useContext(TranslationContext) as TranslationContextType;
    const { request } = useContext(ConnectionContext) as ConnectionContextType;

    const [formError, setFormError] = useState<React.JSX.Element>(() => <HBWError/>);

    const notify = useCallback((error: WidgetErrorType) => {
      // @ts-ignore
      Messenger[error.level](t(error.notificationText, {}, error.notificationText));
    }, [t]);

    const setupFormError = useCallback((error: WidgetErrorType) => {
      setFormError(
        <HBWError
          errorHeader={t(error.header, {}, error.header)}
          errorBody={error.body}
        />,
      );
    }, [t]);

    const addError = useCallback((error: WidgetErrorType) => {
      if (error.level === 'error') {
        setupFormError(error);
      }
      if (error.notificationText) {
        notify(error);
      }
    }, [notify, setupFormError]);

    const safeRequest = (opts: {}) => (
      request(opts)
        .then((response: Response) => {
          if (response.status === 504) {
            addError(CamundaError);
          }
          return response;
        }));

    return (
      <ErrorHandlingContext.Provider value={{
        addError,
        formError,
        safeRequest,
      }}>
        <WrappedComponent {...props} />
      </ErrorHandlingContext.Provider>
    );
  };

  return ErrorHandlingProvider;
};

export default ErrorHandlingContext;
