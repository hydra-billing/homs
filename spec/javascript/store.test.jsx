import React, { useContext } from 'react';
import renderer from 'react-test-renderer';
import StoreContext, { withStoreContext } from 'hbw/components/shared/context/store';
import Dispatcher from 'hbw/dispatcher';
import { withTranslationContext } from 'hbw/components/shared/context/translation';
import { withDispatcherContext } from 'hbw/components/shared/context/dispatcher';
import compose from 'shared/utils/compose';
import { stringify } from 'qs';
import { withConnectionContext } from 'hbw/components/shared/context/connection';
import { waitFor } from '@testing-library/react';

const dispatcher = new Dispatcher();
let emitEvent;
const mockConsumer = {
  subscriptions: {
    create: (
      _,
      received
    ) => {
      emitEvent = async event => received.received(JSON.stringify(event));
    }
  }
};

jest.mock('@rails/actioncable', () => ({ createConsumer: () => mockConsumer }));

const serverURL = 'http://mockURL:3000';
const userIdentifier = 'user@example.com';
const entityClassCode = 'order';
const unassignedTask = { id: 'unassignedTaskID', assignee: null };
const assignedTask = { id: 'assignedTaskID', assignee: 'user@example.com' };
const longFetchedTask = { id: 'longFetchedTask', assignee: 'user@example.com' };

const mockRequest = reqParams => ({
  json: async () => {
    switch (reqParams.url) {
      case `${serverURL}/tasks/${unassignedTask.id}`:
        return unassignedTask;
      case `${serverURL}/tasks/${assignedTask.id}`:
        return assignedTask;
      case `${serverURL}/tasks/${longFetchedTask.id}`:
        await new Promise(_ => setTimeout(_, 50));
        return longFetchedTask;
      case `${serverURL}/translations`:
        return {};
      default:
        return [];
    }
  }
});

const mockConnection = {
  request: mockRequest,
  serverURL
};

const testEvents = {
  create: {
    task_id:        unassignedTask.id,
    event_name:     'create',
    version:        1,
    cache_key:      null,
    assigned_to_me: false
  },
  assignment: {
    task_id:        assignedTask.id,
    event_name:     'assignment',
    version:        2,
    cache_key:      null,
    assigned_to_me: true
  },
  complete: {
    task_id:        assignedTask.id,
    event_name:     'complete',
    version:        3,
    assigned_to_me: true
  },
  delete: {
    task_id:        assignedTask.id,
    event_name:     'delete',
    version:        4,
    assigned_to_me: true
  }
};

describe('<StoreProvider />', () => {
  const TestingComponent = () => {
    const {
      tasks, events, versions, fetching, ready, activeTask
    } = useContext(StoreContext);

    return (
      <div>
        <p className="tasks">{stringify(tasks)}</p>
        <p className="events">{stringify(events)}</p>
        <p className="versions">{stringify(versions)}</p>
        <p className="fetching">{fetching.toString()}</p>
        <p className="ready">{ready.toString()}</p>
        <p className="activeTask">{stringify(activeTask)}</p>
      </div>
    );
  };

  const withContext = compose(
    withDispatcherContext({ dispatcher }),
    withConnectionContext({ connection: mockConnection }),
    withTranslationContext({ locale: { code: 'en' } }),
    withStoreContext({
      userIdentifier,
      entityClassCode,
      widgetURL:         serverURL,
      showNotifications: false,
      dispatcher,
      connection:        mockConnection
    })
  );

  let testComponent;

  beforeEach(async () => {
    const Provider = withContext(TestingComponent);
    await waitFor(() => {
      testComponent = renderer.create(<Provider />);
    });
  });

  afterEach(() => {
    testComponent.unmount();
  });

  it('initializes', async () => {
    expect(testComponent.toJSON()).toMatchSnapshot();
  });

  it('adds task to state on create', async () => {
    await emitEvent(testEvents.create);

    expect(testComponent.toJSON()).toMatchSnapshot();
  });

  it('adds task to state on assign', async () => {
    await emitEvent(testEvents.assignment);

    expect(testComponent.toJSON()).toMatchSnapshot();
  });

  it('removes task from state on complete', async () => {
    await emitEvent(testEvents.assignment);
    await emitEvent(testEvents.complete);

    expect(testComponent.toJSON()).toMatchSnapshot();
  });

  it('removes task from state on delete', async () => {
    await emitEvent(testEvents.assignment);
    await emitEvent(testEvents.delete);

    expect(testComponent.toJSON()).toMatchSnapshot();
  });

  it('ignores task assigned to someone else', async () => {
    await emitEvent({ ...testEvents.assignment, assigned_to_me: false });

    expect(testComponent.toJSON()).toMatchSnapshot();
  });

  it('ignores event with outdated version', async () => {
    const task = longFetchedTask;
    const longCreateEvent = { ...testEvents.create, task_id: task.id };
    const completeEvent = { ...testEvents.complete, task_id: task.id };
    await emitEvent(longCreateEvent);
    await emitEvent(completeEvent);

    expect(testComponent.toJSON()).toMatchSnapshot();
  });
});
