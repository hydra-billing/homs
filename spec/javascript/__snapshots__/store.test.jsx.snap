// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StoreProvider /> adds task to state on assign 1`] = `
<div>
  <p
    className="tasks"
  >
    0%5Bid%5D=assignedTaskID&0%5Bassignee%5D=user%40example.com
  </p>
  <p
    className="events"
  />
  <p
    className="versions"
  >
    assignedTaskID=2
  </p>
  <p
    className="fetching"
  >
    false
  </p>
  <p
    className="ready"
  >
    true
  </p>
  <p
    className="activeTask"
  />
</div>
`;

exports[`<StoreProvider /> adds task to state on create 1`] = `
<div>
  <p
    className="tasks"
  >
    0%5Bid%5D=unassignedTaskID&0%5Bassignee%5D=
  </p>
  <p
    className="events"
  />
  <p
    className="versions"
  >
    unassignedTaskID=1
  </p>
  <p
    className="fetching"
  >
    false
  </p>
  <p
    className="ready"
  >
    true
  </p>
  <p
    className="activeTask"
  />
</div>
`;

exports[`<StoreProvider /> ignores event with outdated version 1`] = `
<div>
  <p
    className="tasks"
  />
  <p
    className="events"
  />
  <p
    className="versions"
  >
    longFetchedTask=3
  </p>
  <p
    className="fetching"
  >
    false
  </p>
  <p
    className="ready"
  >
    true
  </p>
  <p
    className="activeTask"
  />
</div>
`;

exports[`<StoreProvider /> ignores task assigned to someone else 1`] = `
<div>
  <p
    className="tasks"
  />
  <p
    className="events"
  />
  <p
    className="versions"
  >
    assignedTaskID=2
  </p>
  <p
    className="fetching"
  >
    false
  </p>
  <p
    className="ready"
  >
    true
  </p>
  <p
    className="activeTask"
  />
</div>
`;

exports[`<StoreProvider /> initializes 1`] = `
<div>
  <p
    className="tasks"
  />
  <p
    className="events"
  />
  <p
    className="versions"
  />
  <p
    className="fetching"
  >
    false
  </p>
  <p
    className="ready"
  >
    true
  </p>
  <p
    className="activeTask"
  />
</div>
`;

exports[`<StoreProvider /> removes task from state on complete 1`] = `
<div>
  <p
    className="tasks"
  />
  <p
    className="events"
  />
  <p
    className="versions"
  >
    assignedTaskID=3
  </p>
  <p
    className="fetching"
  >
    false
  </p>
  <p
    className="ready"
  >
    true
  </p>
  <p
    className="activeTask"
  />
</div>
`;

exports[`<StoreProvider /> removes task from state on delete 1`] = `
<div>
  <p
    className="tasks"
  />
  <p
    className="events"
  />
  <p
    className="versions"
  >
    assignedTaskID=4
  </p>
  <p
    className="fetching"
  >
    false
  </p>
  <p
    className="ready"
  >
    true
  </p>
  <p
    className="activeTask"
  />
</div>
`;
